randovania/games/dread/layout/dread_damage_state.py

```diff
@@ -0,0 +1,112 @@
+
+from typing import TYPE_CHECKING, Self, override
+
+from randovania.game_description.requirements.resource_requirement import ResourceRequirement
+from randovania.resolver.energy_tank_damage_state import EnergyTankDamageState
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
+
+    from randovania.game_description.resources.item_resource_info import ItemResourceInfo
+    from randovania.game_description.resources.resource_collection import ResourceCollection
+
+
+class DreadDamageState(EnergyTankDamageState):
+    _use_immediate_energy_parts: bool
+    _energy_part_item: ItemResourceInfo
+
+    def __init__(
+        self,
+        starting_energy: int,
+        energy_per_tank: int,
+        energy_tank: ItemResourceInfo,
+        use_immediate_energy_parts: bool,
+        energy_part_item: ItemResourceInfo,
+    ):
+        super().__init__(starting_energy, energy_per_tank, energy_tank)
+        self._use_immediate_energy_parts = use_immediate_energy_parts
+        self._energy_part_item = energy_part_item
+
+    def __copy__(self) -> Self:
+        return self._duplicate()
+
+    @override
+    def resources_for_energy(self) -> Generator[ItemResourceInfo]:
+        yield self._energy_tank
+        yield self._energy_part_item
+
+    def _maximum_energy(self, resources: ResourceCollection) -> int:
+        num_tanks = resources[self._energy_tank]
+        num_parts = resources[self._energy_part_item]
+        num_whole_tanks = num_tanks + num_parts // 4
+        num_partial_parts = num_parts % 4
+        energy_from_partial = (self._energy_per_tank * num_partial_parts) // 4
+
+        return self._starting_energy + (self._energy_per_tank * num_whole_tanks) + energy_from_partial
+
+    def _duplicate(self) -> Self:
+        result = DreadDamageState(
+            self._starting_energy,
+            self._energy_per_tank,
+            self._energy_tank,
+            self._use_immediate_energy_parts,
+            self._energy_part_item,
+        )
+        result._energy = self._energy
+        return result
+
+    def _energy_part_difference(
+        self,
+        new_resources: ResourceCollection,
+        old_resources: ResourceCollection,
+    ) -> int:
+        return new_resources[self._energy_part_item] - old_resources[self._energy_part_item]
+
+    @override
+    def apply_collected_resource_difference(
+        self, new_resources: ResourceCollection, old_resources: ResourceCollection
+    ) -> Self:
+        if self._energy_tank_difference(new_resources, old_resources) > 0:
+            return self._at_maximum_energy(new_resources)
+        elif (new_parts := self._energy_part_difference(new_resources, old_resources)) > 0:
+            if self._use_immediate_energy_parts:
+                result = self._duplicate()
+                result._energy = (
+                    self._energy
+                    + (self._energy_per_tank * new_parts // 4)
+                    + (self._energy_per_tank * (new_parts % 4)) % 4
```

Comment:
Dunno what I'm looking at here xD, but if this is calculating a new energy value, this needs to be `+ (self._energy_per_tank * (new_parts % 4)) / 4` instead of the modulo operator at the end.
E.g. `new_parts = 2` => `(100 * (2 % 4)) % 4` => `(100 * 2) % 4` => `200 % 4` => `0` but it should be `200 / 4` => `50`

Reply:
I suspected that there would be a more succinct way to write this, but I I ended up with this when I started to consider numbers like 4 and 5.

Reply:
Just `self._energy_per_tank / 4 * new_parts` ? Would always be just `25 * new_parts`

Reply:
The other thing I wanted to take into some kind of consideration was what happens when you don't have a multiple of 4 energy per tank. But I'm not sure what the game would do in those cases.

Reply:
I ended up simplifying this by writing it in terms of `_maximum_energy`.